code_generator/GdkCodeGenerator/Templates/UnityComponentConversionGenerator.tt

```diff
@@ -9,493 +9,333 @@
                 {
+                    LogDispatcher.HandleLog(LogType.Error, new LogEvent(CommandIndexNotFound)
+                        .WithField(LoggingUtils.LoggerName, LoggerName)
+                        .WithField(LoggingUtils.EntityId, op.EntityId.Id)
+                        .WithField("Command Index", commandIndex)
+                        .WithField("Component Type", "<#= componentDetails.TypeName #>")
+                    );
                     return;
                 }
 
-                var unityResponsePayload = op.Response.HasValue 
-                    ? <#= commandDetails.ResponseType #>.ToNative(op.Response.Value.Get().Value) 
-                    : (<#= commandDetails.ResponseType #>?) null;
-                var outgoingRequest = (<#= commandDetails.CommandName #>.OutgoingRequest) requestContext.Request;
-                var response = new <#= commandDetails.CommandName #>.Response(op.EntityId.Id, op.Message, (CommandStatusCode) op.StatusCode, unityResponsePayload, outgoingRequest.RawRequest);
-
-                view.AddCommandResponse(entity, response, <#= commandDetails.CommandName #>ResponsePool);
+                CommandResponses[commandIndex](op);
             }
 
-            private void Send<#= commandDetails.CommandName #>Request(Connection connection, <#= commandDetails.CommandName #>.OutgoingRequest outgoingRequest)
+            private void ApplyAuthorityChange(Unity.Entities.Entity entity, Authority authority, long entityId)
             {
-                var requestPayload = <#= commandDetails.RequestType #>.ToSpatial(outgoingRequest.RawRequest);
-                var request = new <#= componentDetails.FullyQualifiedSpatialTypeName #>.Commands.<#= commandDetails.CommandName #>.Request(requestPayload);
-
-                var requestId = connection.SendCommandRequest(new global::Improbable.EntityId(outgoingRequest.TargetEntityId), request, null);
+                switch (authority)
+                {
+                    case Authority.Authoritative:
+                        if (!entityManager.HasComponent<NotAuthoritative<<#= componentDetails.TypeName #>>>(entity)) 
+                        {
+                            LogDispatcher.HandleLog(LogType.Error, new LogEvent(InvalidAuthorityChange)
+                                .WithField(LoggingUtils.LoggerName, LoggerName)
+                                .WithField(LoggingUtils.EntityId, entityId)
+                                .WithField("New State", Authority.Authoritative)
+                                .WithField("Expected Old State", Authority.NotAuthoritative)
+                            );
+                            return;
+                        }
 
-                RequestIdToRequestContext.Add(requestId.Id, new RequestContext(outgoingRequest.SenderEntityId, outgoingRequest));
-            }
+                        entityManager.RemoveComponent<NotAuthoritative<<#= componentDetails.TypeName #>>>(entity);
```

Comment:
If we don't do the explicit checks it will fail here, which in my mind seems good enough. I don't know why a user would remove a component themselves, but if they do this will tell you what you did. I'd prefer we add validation for other events later, rather than while we are still prototyping stuff.

Reply:
The problem with letting the failure occur in the ECS is that it doesn't surface very helpful messages - but we can remove the handling from here if you prefer

Reply:
Yes I think so

Reply:
Actually - we can refactor the logging call into a utility method, but if we refactor the entire thing you get something like this anyway:

```
if (!CheckEntityExists(entityId, "op type", out var entity))
{
	return;
}
```